### PR TITLE
23157 notes

### PR DESCRIPTION
**Pull request by jnewbery**

### CODE REVIEW

**Comment by jnewbery** [comment]
@glozow revoo plz

File: _posts/2021-10-06-#23157.md [hunk 1]
```diff
@@ -10,15 +10,80 @@ status: upcoming
...
+  optimized for various operations, such as constructing blocks.
+
+- The mempool must only contain non-conflicting, valid transactions which
+  could theoretically be included in a future block. There are therefore
+  many [invariants](https://en.wikipedia.org/wiki/Class_invariant) such as:
+
+  - a transaction may only appear in the mempool if all of its inputs are
+    either in the UTXO set or are the outputs of other mempool transactions
+  - no two transactions in the mempool may spend the same output
+  - the {mempool|ancestor} {fees|count|size} cached for the transaction must
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:33]: I assume this is what it's supposed to say? Or self | descendant | ancestor?
```suggestion
  - the {descendant|ancestor} {fees|count|size} cached for the transaction must
```

File: _posts/2021-10-06-#23157.md [hunk 2]
```diff
@@ -10,15 +10,80 @@ status: upcoming
 commit:
 ---
 
-_Notes and questions to follow soon!_
+We recently [looked at PR 22677](./22677), which cuts the circular dependency
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:13]: I usually put the # sign in front of PR numbers but up to you
```suggestion
We recently [looked at PR #22677](./22677), which cuts the circular dependency
```

File: _posts/2021-10-06-#23157.md [hunk 3]
```diff
@@ -10,15 +10,80 @@ status: upcoming
...
+between validation and txmempool. This week's PR was split off from that PR,
+taking a slightly different approach to simplify the `CTxMemPool::check()`
+function.
 
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+- The `CTxMemPool` class is a data structure that stores the set of unconfirmed
+  transactions, along with various methods to update or query that set. As well
+  as the transactions themselves, the mempool stores metadata about the
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:22]:
```suggestion
  transactions, along with various methods to update or query that set. In addition
  to the transactions themselves, the mempool stores metadata about the
```

File: _posts/2021-10-06-#23157.md [hunk 4]
```diff
@@ -10,15 +10,80 @@ status: upcoming
...
+4. What does the `UpdateCoins()` function do? What is it replaced with in
+   this PR?
+
+5. What was `waitingOnDependants` used for before this PR? Explain what
+   [this while
+loop](https://github.com/bitcoin/bitcoin/blob/446b70669/src/txmempool.cpp#L779-L790) was doing. Why can it now be removed?
+
+6. How does the `GetSortedDepthAndScore()` function work? How does it ensure
+   that the vector that it returns doesn't contain any descendants before
+   their ancestors?
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:82]:
```suggestion
6. How does the `GetSortedDepthAndScore()` function work? How does sorting
    by the number of ancestors ensure that ancestors precede descendants?
```

File: _posts/2021-10-06-#23157.md [hunk 5]
```diff
@@ -10,15 +10,80 @@ status: upcoming
...
+
 ## Questions
+
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
--->
 
+2. Did you run the bench tests before and after the changes to
+   `CTxMemPool::check()`? Were your results the same as reported in the PR?
+
+3. Is it important for these checks to be fast? Why?
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:71]:
```suggestion
3. Is it important for these checks to be fast? Why or why not?
```

File: _posts/2021-10-06-#23157.md [hunk 6]
```diff
@@ -10,15 +10,80 @@ status: upcoming
...
+  - no two transactions in the mempool may spend the same output
+  - the {mempool|ancestor} {fees|count|size} cached for the transaction must
+    be the correct value
+
+- There are several public methods that can alter the contents of the
+  mempool. Those public methods can be called in many different
+  circumstances, such as:
+
+  - removing transactions when a block is connected
+  - expiring old transactions or removing transactions when limiting the
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:41]:
```suggestion
  - expiring old transactions or evicting transactions by feerate when limiting the
```

File: _posts/2021-10-06-#23157.md [hunk 7]
```diff
@@ -10,15 +10,80 @@ status: upcoming
...
 ## Notes
 
+- The `CTxMemPool` class is a data structure that stores the set of unconfirmed
+  transactions, along with various methods to update or query that set. As well
+  as the transactions themselves, the mempool stores metadata about the
+  transactions (such as ancestor/descendant dependencies) that makes it highly
+  optimized for various operations, such as constructing blocks.
+
+- The mempool must only contain non-conflicting, valid transactions which
+  could theoretically be included in a future block. There are therefore
```
Comment (glozow) [inline _posts/2021-10-06-#23157.md:27]:
```suggestion
  could theoretically be included in the next block. There are therefore
```

**Review by glozow — COMMENTED** [review]
nice

File: _posts/2021-10-06-#23157.md (hunk 7 above)
Comment (glozow) [reply _posts/2021-10-06-#23157.md:27]: (stricter than future block. must be minable immediately)

File: _posts/2021-10-06-#23157.md (hunk 1 above)
Comment (jnewbery) [reply _posts/2021-10-06-#23157.md:33]: oops. Thank you!

File: _posts/2021-10-06-#23157.md (hunk 2 above)
Comment (jnewbery) [reply _posts/2021-10-06-#23157.md:13]: Done!

File: _posts/2021-10-06-#23157.md (hunk 4 above)
Comment (jnewbery) [reply _posts/2021-10-06-#23157.md:82]: I think this gives away too much of the answer. I've changed the wording of the second part of the question though: "Why does the returned vector not contain any descendants before their ancestors?"

File: _posts/2021-10-06-#23157.md (hunk 6 above)
Comment (jnewbery) [reply _posts/2021-10-06-#23157.md:41]: Done

File: _posts/2021-10-06-#23157.md (hunk 7 above)
Comment (jnewbery) [reply _posts/2021-10-06-#23157.md:27]: You're right!

**Comment by jnewbery** [comment]
Thanks for the review @glozow. I've addressed all of your review comments.

**Pull request merged**